### PR TITLE
fix: drain and close API response bodies

### LINKED ISSUE #316: Resources leak through net/http transport goroutines
## Summary
Hi! I found that there is a resource leak problem in case of lot of events. The amount of goroutines, used fd and consumed memory is growing and does not try to decrease.
In my case I use sentry-go in logrus hook plugin and got thousands goroutines and fd limit in case of thousands of events.

For each transport flush there hung a couple "zombie"-routines
net/http.(*persistConn).readLoop
net/http.(*persistConn).writeLoop

## Steps To Reproduce

Any code with big amount events to flush.
https://github.com/onrik/logrus/blob/master/sentry/sentry.go

## Expected Behavior

Http.Client using keep-alive connections.

## Screenshots

image

### SDK
 - `sentry-go` version: 0.9.0
 - Go version: 1.15
 - Using Go Modules? yes

### PR DESCRIPTION
**Pull request by rhcarvalho**
This fixes the two transport implementations to drain (up to a safe limit) and close the response body from API calls, allowing the net/http client to reuse connections.

Fixes #316.
Closes #317 (replaces it).

Thanks @laoleesch.

### CODE REVIEW

File: transport.go [hunk 1]
```diff
@@ -17,6 +19,17 @@ const defaultBufferSize = 30
...
+// maxDrainResponseBytes is the maximum number of bytes that transport
+// implementations will read from response bodies when draining them.
+//
+// Sentry's ingestion API responses are typically short and the SDK doesn't need
+// the contents of the response body. However, the net/http HTTP client requires
+// response bodies to be fully drained (and closed) for TCP keep-alive to work.
+//
+// maxDrainResponseBytes strikes a balance between reading too much data (if the
+// server is misbehaving) and reusing TCP connections.
+const maxDrainResponseBytes = 16 << 10
```
Comment (kamilogorek) [inline transport.go:31]: Not sure why `16 << 10` instead of `1024 * 16` or just `16384`, but ok :p

File: transport.go [hunk 2]
```diff
@@ -472,15 +488,18 @@ func (t *HTTPSyncTransport) SendEvent(event *Event) {
 	)
 
 	response, err := t.client.Do(request)
-
 	if err != nil {
```
Comment (kamilogorek) [inline transport.go:491]: I assume that `err` and `response` are mutually exclusive, and only one of them can be `nil` at the time?

File: transport_test.go [hunk 3]
```diff
@@ -361,3 +363,102 @@ func TestHTTPTransport(t *testing.T) {
...
+
+	// If server responses are too large, the SDK should close the
+	// connection instead of consuming an arbitrarily large number of bytes.
+	largeResponse = true
+
+	// Next event, first one to get a large response, reuses the connection.
+	tr.SendEvent(event)
+	checkLastConnReuse(true)
+
+	// All future events create a new TCP connection.
```
Comment (kamilogorek) [inline transport_test.go:450]: Should they all create new TCP connections, or only the first one after the previous one was closed (due to too large drainage), and remaining 9 should reuse it?

Comment (kamilogorek) [reply transport_test.go:450]: Ah, `largeResponse` is still kept `true` here, so they all should use new I guess.

**Review by kamilogorek — APPROVED** [review]

File: transport.go (hunk 2 above)
Comment (rhcarvalho) [reply transport.go:491]: https://golang.org/pkg/net/http/#Client.Do

>  On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails [...]

File: transport_test.go (hunk 3 above)
Comment (rhcarvalho) [reply transport_test.go:450]: Yes. The tests cover two types of server behavior, first a server that behaves like current-times Relay, and second simulates sending more data, more than we're willing to read and discard. In the latter case, keep-alive should not work unless the Go HTTP client changes in a future release.

**Pull request merged**